### DISCUSSION TITLE
How to make webrtc penetration in the internet

### DISCUSSION THREAD

**Discussion by fayfive** [pion/webrtc#1816] [Q&A]
version:3.0.29
system:ubunbtu16.04

I have a camera And I using a PC to convert the camera signal to webrtc.
The stun is URLs: []string{"stun:stun.l.google.com:19302"}.
In the LAN very thing is OK. So I decide to make it in the internet.
The method is that:
1. I map a http port in the internet as a control signaling server.
2. I visit the web, the web will creatoffer and send the IceCandidateInit(s) and sdp to the server. When the server receive the http request, it will SetRemot and AddIceCandidate() .At last the serverwill creatanswer and send the IceCandidateInit(s) answer sdp to the client web.  The web will SetRemot and AddIceCandidate. Through wireshark, I can see web stun binding request and the google binding Sucess Response. And then the web stun the server IPs. I can see one of IPs is the internet IP of the server and the port is what I get form the OnIceCandidate().  A moment latter, webrtc server show that OnICEConnectionStateChange connectionState.String() is failed

Why the webrtc server doesn't response the client?

**Comment by Sean-Der** [comment]
It is possible that you are behind a NAT that doesn't support address independent mapping/filtering. [WebRTC for the Curious#Networking](https://webrtcforthecurious.com/docs/03-connecting/) has a section explaining more. [stun-nat-behaviour](https://github.com/pion/stun/tree/master/cmd/stun-nat-behaviour) is a good tool to learn more also.

I would run a TURN server and make it acessiable via TCP and UDP to support all networks. You could run your own or using a provider like Twillio.

**Reply by fayfive** [reply]
Thanks for your answer. I don’t know the TURN server. Is there any TURN that is open source to recommend to me? It’s the best writing in golang 😬

**Reply by Sean-Der** [reply]
Yes! https://github.com/pion/turn is one. You also could use https://github.com/gortc/gortcd

The most popular is https://github.com/coturn/coturn

**Reply by fayfive** [reply]
Thanks a lot!
It works well again. When I deployed the coturn.
After that I will learn the TURN and write the code base on pion/turn.^_^
